Support for POSIX signal

ekrich opened the pull request:
This is separated from the `aio.h` PR. Doing both is a little too much so will complete this one first.

coreyoconnor commented on posixlib/src/main/scala/scala/scalanative/posix/signal.scala:
```diff
@@ -0,0 +1,468 @@
+ */
+@extern
+object signal {
+  // define the following macros, which shall expand to constant expressions with distinct values
```
Is the reference to "macros" here a TODO? Looks like this code does not currently use macros. But could be replaced with a macro?

ekrich replied:
We have to have helper functions in C to support macros currently. The comment is from the standard and that portion is complete.
http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/signal.h.html

coreyoconnor replied:
Ah! Makes sense. I see now the reference is to the C preprocessor macros used to define the constant. I was stucking thinking only about Scala macros. LGTM!

ekrich commented:
I updated this PR so if someone wanted to finish it they could or I could finish it if someone really wants it.

Originally I thought `aio` might be useful and `signal` is required. In the mean time I read that `aio` is not that useful. There is some `aio` code there but only because I just didn't want to blindly remove it.

ekrich commented:
Since @LeeTibbert wants to use this PR, I have it to a state that it could be merged knowing that it is not 100% complete. This has been a large effort just to get to this state. What does this PR include:

1. Includes all of the Scala interface file.
2. Moves and deprecates items in `clib` that should have been in `posixlib`.
3. Includes all the C macro code.
4. Includes the `implicit`s for the `struct`s and `union`s.

It doesn't include the C code needed for copying Scala structs to C structs but a large part of the API can be used as included in this PR, along with the `clib` signal functions and macros which complete the POSIX API.

LeeTibbert commented:
@ekrich Thank you for promoting this work to a PR.  It gives me `siginfo_t` which will
make completing  PR #2157 easier/possible.   PR #2157 will, in turn, will ease other
work, such as a more robust test-interface TestMain for debugging intermittent errors
on macOS.   `siginfo_t` need not work, nor be right, it just must exists. Of course,
working would be a bonus.

Reviewers: 
I concur that signal support is the work of ages and will need generations & evolutions.
This evolution, albeit incomplete, is cost effective; it provides scaffolding.  The next
generation probably is not cost effective at this point.

ekrich commented:
This will probably need to be rebased on top of the FreeBSD if this gets merged. https://github.com/scala-native/scala-native/pull/2148

I believe that BSD may come up short on the full POSIX implementation so this PR blocked by the conditional PR. https://github.com/scala-native/scala-native/pull/2206

WojciechMazur commented on posixlib/src/main/resources/scala-native/signal.h:
```diff
@@ -0,0 +1,6 @@
+#ifndef SIGNAL_H
```
Looks like this header is not needed right now, it can probably be removed.

ekrich replied:
Originally added for `aio` so removed for now.

WojciechMazur commented on posixlib/src/main/resources/scala-native/aio.c:
```diff
@@ -0,0 +1,65 @@
+// scala native functions
+
+int scalanative_aio_cancel(int fd, struct scalanative_aiocb *sn_aiocb) {
+    return 0;
```
As I understand this and following functions are only a stubs. Are they used anywhere? Is it not dangerous to provide functions that be used by user, which are actually not doing anything and don't report they're stubs?    We should either remove them or make them  crash the program.

ekrich replied:
Removed, `aio` code slipped through from a former draft of this PR.

WojciechMazur commented on posixlib/src/main/scala/scala/scalanative/posix/aio.scala:
```diff
@@ -0,0 +1,127 @@
+object aioOps {
+  import aio._
+
+  // TODO: move to signal
```
Two remaining todos in this file. Is there something that does not allow us to move it to signal?

ekrich replied:
Gone.

WojciechMazur commented on posixlib/src/main/scala/scala/scalanative/posix/signal.scala:
```diff
@@ -0,0 +1,407 @@
+
+  // Note 1: Linux
+  //  @name("scalanative_sig_hold")
+  //  def SIG_HOLD: CFunctionPtr1[CInt, Unit] = extern
```
Shouldn't this be uncommented? Same applies to code under another notes

ekrich replied:
I cleaned these up and put a good comment at the top explaining. Basically, these are in the spec but don't work on either Linux or macOS. They are left in the code so a future developer will not go chasing these down. Comparing the spec to the code is all visual and then parts are in the portable `clib` so merging all these things together is really hard.

WojciechMazur reviewed (changes requested):
Looks fine, I understand that it does not provide full support yet, however we can try to make it a bit more safe right now, especially in terms of unimplemented aio methods. We should also move signalOps defined in aio to signal if that is possible before merge.

ekrich commented:
Thanks you for the review. I didn't even realize I was pushing AIO changes. I'll have to look through this more carefully and will respond to the review comments.

LeeTibbert commented on clib/src/main/resources/scala-native/signal.c:
```diff
@@ -19,5 +19,3 @@ int scalanative_sigint() { return SIGINT; }
 
 int scalanative_sigterm() { return SIGTERM; }
-
-int scalanative_sigusr1() { return SIGUSR1; }
```
I have not run a program, but I think that deleting line 23 here
will cause lines 41 & 42 in signal.scala to have a
"Missing symbol" at link time.

I believe the intended change is to remove this line
when the deprecation grace period of lines 41/32 in signal.scala
expires and those lines are removed.

ekrich replied:
The linkage between the `.c.o` and `.ll.o`(which contains the `extern`s) happens at link time so all the C object code is essentially a big blob with no namspace. This is why we can only have the C code in one place.

I would agree if we omitted `posixlib` for Windows compiling for example. I am hoping we can just remove the Scala interface in `clib` that is deprecated in the `0.4.2` release. I don't think a long deprecation cycle is warranted based on fact that we have very little usage. I also suspect or at least hope that we can have some minor releases in the mean time before we officially change things for Windows. If we do change sooner then the code would compile but fail to link which is like an error on deprecation but only on the Windows platform that shouldn't have the POSIX constant anyway.

You can move a C file from one project to another and see that it still works to satisfy your curiosity.

LeeTibbert replied:
Given the information in the reply just above and further study, I think the problem statement
for this point is:

```
 SIGUSR1 is a Gnu C extension. It is not  part of the 
`EEE Std 1003.1-2017 (Revision of IEEE Std 1003.1-2008)` specification, so it should
not be included in clib.  It is part of clib in SN 0.4.0 because of the impure initial
seeding of clib.

The presence of this declaration prevent compilation on Windows and possibly 
other non-GnuC systems, so it should be surgically removed after a short deprecation
cycle on Gnu C Linux.
```

Given that problem statement, or a close evolution of it, something similar to the
following would allow compilation on Windows (and non-GnuC systems) yet
not introduce a decidedly unfriendly link time "Missing Symbol"  error.
(By my understanding, "Deprecated" means "continues to work for a short time"
not a link time error.)

```
// not compiled, for concept
#ifdef __GNUC__
// Remove after deprecated & only caller in posix signal.scala is removed.
int scalanative_sigusr1() { return SIGUSR1; }
#endif
```

The technique is general to deprecating & removing other GnuC extensions in clib.

ekrich replied:
This is a fine idea but that is why I moved it from `clib` to `posixlib`. This way windows will not see it as explained above. I think if we get to a point were this is not removed before Win support is added we can deal with it then. My guess is Win support is a ways off.

LeeTibbert replied:
OK, now I see! Duh!  This is basically a (well hidden) move from clib to posixlib.  Searching again,
I see SIGUSR1 in the 2018 Open Group Posix spec.

A good change overall.   

I got lost in the steps to the dance. When I saw the deprecation code in
clib signal.scala I expected, according to the current orginization, for the
implementation corresponding to `@name scalanative_sigusr1` to be
in clib signal.c.   

Yes, it works due to C's flat namespace and our conventions, but it
is hard to follow the elegant intervention.

Perhaps adding a comment to the clib/libc signal.scala. Someday,
someone will sing your praises.

```
@name("scalanative_sigusr1") // implemented in posix signal.c
```

LeeTibbert commented on clib/src/main/scala/scala/scalanative/libc/signal.scala:
```diff
@@ -7,7 +7,9 @@ import scalanative.unsafe._
   // Signals
-
+  @deprecated(
+    "Not in C standard, use posix signal. To be removed in next version",
```
re: "Not in C standard".  Which C standard C89, C11, ISO C?
There are any number of C standards in play.

I think you mean "IEEE Std 1003.1™-2017 (Revision of IEEE Std 1003.1-2008)". You/we probably only want to give the portion to
the left of the open parenthesis. 

Yes, picky, but I think lack of precision here is an indicator of 
`clib` not having a `00_READ_ME` which gives a clear statement
of which standard files in that project are to follow.   Not to try to 
solve the general `clib` problem here, but I think there is an 
opportunity to remove some technical debt from `clib`.

ekrich replied:
The function it not in any C standard - I think the intent here is just let the deprecation point the user to the correct place, This is really only for a UNIX user who may have used this function from `clib`. I can change the deprecation warning and I probably should say what version it will be removed in or deprecated in.

ekrich replied:
Ok, I updated those too, hopefully to your liking.

LeeTibbert commented on javalib/src/main/scala/java/lang/UnixProcess.scala:
```diff
@@ -6,10 +6,17 @@ import java.util.concurrent.TimeUnit
 import scala.scalanative.unsafe._
-import scala.scalanative.libc.{errno => err, signal => sig, _}
-import sig._
+import scala.scalanative.libc.{errno => err}
```
This is a second generation, picky mode, observation, not meant
to hold up the merge of the PR.

The distinction between the errno intended by 'err` and `e`
is pretty fine.  Given the path that got us to having two subtlety 
different errnos, I suggest making it easier for future readers
to determine, at the point of reading,  which one is meant. 

Something like:
'```
import scala.scalanative.libc.{errno => libcErrno}
import scala.scalanative.posix.{
  fcntl,
  pthread,
  signal,
  sys,
  time,
  unistd,
  errno => posixErrno
}
```
Eric, If you agree with the concept but do not have time/energy to make
the fix in this PR.  I can alway submit a followup PR.  Let he/she who proposes, dispose (An old Digital Equipment Corporation saying/principal)
to keep work on a given item from expanding without end and consuming
the person doing the original work).

ekrich replied:
I am just removing the `signal` import and not changing the other anything else. I think that is the correct action.

ekrich replied:
I think working the errno is out of scope. Maybe they should be `cerr` and `perr` but without the extra imports to bring more things into scope.

LeeTibbert replied:
Understood & agreed; minimal changes and those for only the current topic.

I will add UnixProcess.scala to the list of places I hope to never visit....

LeeTibbert commented on javalib/src/main/scala/java/lang/UnixProcess.scala:
```diff
@@ -26,7 +33,7 @@ private[lang] class UnixProcess private (
     errfds: Ptr[CInt]
 ) extends Process {
-  override def destroy(): Unit = kill(pid, 9)
+  override def destroy(): Unit = signal.kill(pid, 9)
```
Another pick review mode item:

Granted that anybody reading this code should understand that signal 9
means SIGKILL, using Posix signal.SIGKILL (included already) would
remove a magic number.

Same offer, if you agree with concept but want me to do the work in a further
PR, call out.

ekrich replied:
No I will fix that.

LeeTibbert commented on posixlib/src/main/scala/scala/scalanative/posix/signal.scala:
```diff
@@ -0,0 +1,411 @@
+  // that have a type compatible with the second argument to, and the return value of, the signal() function,
+  // and whose values shall compare unequal to the address of any declarable function.
+
+  // Note 1: Linux
```
Having the clue is good, but I must confess that I can not de-code it.
Is it trying to say that the XSI obsolete SIG_HOLD is defined 
only on Linux? or not on Linux?  

As someone who has spent __lots__ of time in understandably
incomplete SN posix header files I appreciate having the 
comment marker for entry points in the standard. They let
me know that the entry point was considered and is 
intentionally not implemented.

The same comment convention is used in other places.

ekrich replied:
See this - was updated.
https://github.com/scala-native/scala-native/pull/1362/files/54a95930e67d8824aa17a5375fc3e7244d29cb79#diff-dc2e6a694318d6132b77e0d77932935258be9cda1c6c2d929cd50e8e96054f8eR16-R18

LeeTibbert commented on posixlib/src/main/scala/scala/scalanative/posix/signal.scala:
```diff
@@ -0,0 +1,411 @@
+  type timespec = time.timespec
+
+  // define the following data types
+  type sig_atomic_t = CInt
```
Granted all the discussions about SN, uni-threading, and multithreading.
Granted also that `concurrent/atomic` belies it name in the multighread
case.

Since this is posixlib, please consider using `javalib/src/main/scala/java/util/concurrent/atomic/AtomicInteger.scala`.

ekrich replied:
They are all just `int`s so I am guess this is correct or possible generated.
>sig_atomic_t
    Possibly volatile-qualified integer type of an object that can be accessed as an atomic entity, even in the presence of asynchronous interrupts.

LeeTibbert commented on posixlib/src/main/scala/scala/scalanative/posix/signal.scala:
```diff
@@ -0,0 +1,411 @@
+  type pid_t          = types.pid_t
+  type pthread_attr_t = types.pthread_attr_t
+
+  type sigevent = CStruct5[
```
This comment can be skipped if one does not want to go down
the rabbit hole.   Understood entirely that this is an area of complexity,
which is probably going to cause unending grief no matter how it is done.

We may both have other, urgent, areas of greater interest and with
a better cost/benefit ratio.

Since this is review mode, Here we go, after the White Rabbit...

Lines 65/66 define the type sigval.  That definition appears reasonable
but may change as it is exercised in the field.   For now lets go with it.

If those lines were moved to jujst before here,  (yes, I am old Pascal style, declare before use. Understood they they do not technically need to be moved), I believe the type `sigval` could be used in the `sigevent` notification 
rather than as `Ptr[sigval]`.  That is,  use the type rather than a Ptr
to the type.  Since we are doing `copy-in/copy-out` semantics anyway,
the underlying OS will not be writing directly into the SN structure.

IMO, removing the level of indirection (if possible) make it easier to 
trace the code.

ekrich replied:
No, this has to be a `Ptr` and de-referenced in the C helper.

LeeTibbert commented on posixlib/src/main/scala/scala/scalanative/posix/signal.scala:
```diff
@@ -0,0 +1,411 @@
+    CInt,       // si_errno If non-zero, an errno value associated with this signal, as described in <errno.h>
+    pid_t,      // si_pid Sending process ID
+    uid_t,      // si_uid Real user ID of sending process
+    Ptr[Byte],  // void *si_addr Address of faulting instruction (Ptr instead - is void * a fcn?)
```
Sorry, I do not understand the comment.
Perhaps something like 
```
Ptr[Byte], // void *si_addr Address of faulting instruction
```
One could add text that Ptr[Byte] is the SN idiom for 'void *', but
I think the terse declaration is already saying that (Of course, I have,
or pretend to have, a prior knowledge base containing that idiom.)

ekrich replied:
This is exactly from the documentation - the C code is left in the docs because a pointer to void translates to `Ptr[Byte]` in Scala Native. This is also the case for the platform specific structs. They could be `Ptr[Byte]` or a `CArray` if you know the max size.

You can click on Headers/signal you can follow along with the docs. The code is in the order of the docs to try and make things easier to review or update. https://pubs.opengroup.org/onlinepubs/9699919799/

ekrich replied:
ok got it.

LeeTibbert commented on posixlib/src/main/scala/scala/scalanative/posix/signal.scala:
```diff
@@ -0,0 +1,411 @@
+    uid_t,      // si_uid Real user ID of sending process
+    Ptr[Byte],  // void *si_addr Address of faulting instruction (Ptr instead - is void * a fcn?)
+    CInt,       // si_status Exit value or signal
+    CLong,      // si_band Band event for SIGPOLL
```
Since si_band is `OB_XSR` (Obsolete, XSI Streams) in the spec, and we are
using copy-in/copy-out semantics, I suggest this field be changed to a placeholder comment giving the name and that it is obsolete. That would
reduce complexity and cost spent on an obsolete field.  If done, a 
corresponding change to the underlying C code and perhaps any Ops
implicit.

Again, if you agree with the concept but do not have time/patience for the
change, let me know.

ekrich replied:
I guess we could do this but we haven't in the past - we don't have any package docs to hold stuff like that.

ekrich replied:
Not much in the way of docs - is the plan or do we want people to go the spec and stuff online? https://www.javadoc.io/doc/org.scala-native/posixlib_native0.4_2.11/latest/index.html#package

LeeTibbert replied:
I never knew or had see `www.javadoc.io/org.scala-native/posixlib_native`.  Learn something
every day.

I am usually a "document it all & do it early" person but my thinking is that at most this
should be documented as `experimental`.  I think it needs/deserves some more maturity
before even that.

LeeTibbert reviewed (approved):
Eric, you did an enormous amount of work, thank you.
And thank you for the opportunity to review it.

At the 8.000 meter level, this looks good to me - LGTM (as an
initial implementation, suitable for bootstrapping, and 
where we all know that it will need evolution.)

This PR, as is, gives me what I need to peck away
at implementing `wait.h` (see PR #2157) in my spare time.

I left some review comments at the 1000 meter level.
If folks agree with the suggested changes, I propose that
I make them in a series of PRs after this one is merged.

LeeTibbert commented:
Looks even better after 2021-04-06 changes. They remove some sharp edges.

ekrich commented:
Perhaps we should have some basic documentation.

LeeTibbert commented:
Perhaps  listing signal in the table in the "normal" place `docs./lib/javalib.rst` 
with a footnote in the listing which directs attention to a short paragraph there which states that signal
exists but is highly EXPERIMENTAL and contents are subject to change without deprecation periods.
 If I understand correctly it could describe Ops and helper methods (the Ops themselves) and 
recommend their use in order to deal with `Ptr[sigval]` (and one other?) correctly & easily.

I think if I were reading the doc looking for signal, I would find that and understand the hazard
if I decided to proceed past the EXPERIMENTAL notice.

Thoughts?

WojciechMazur commented on posixlib/src/main/scala/scala/scalanative/posix/signal.scala:
```diff
@@ -0,0 +1,411 @@
+  type ucontext_t = CStruct4[
+    Ptr[Byte], // ucontext_t *uc_link Ptr to the context that is resumed when this context returns (Ptr instead)
+    sigset_t,  // c_sigmask The set of signals that are blocked when this context is active
+    Ptr[
+      stack_t
+    ],         // uc_stack The stack used by this context (Ptr instead of value)
```
Was such formatting done by scalafmt? Looks a bit weird. 
Anyway, I think it looks good overall. Can you please rebase it onto the current master to make sure that it passes check-lint in the CI?

ekrich replied:
It has to do with the alignment being `most` trying to align thing vertically and then the comment was too long. I shortened the comment enough to make it work for now. We should discuss trying to get scalafmt upgraded at some point but we will need to change some settings and accept some changes.

ekrich replied:
I guess I missed this one on the prior review, sorry.

ekrich commented:
@LeeTibbert I think I addressed all of your concerns. Any that may persist could be corrected if anyone wants to attempt to use this and add any struct copying code.

@WojciechMazur This also moves the errant `SIGUSR1` from `clib` to `posixlib` as it should be. It would be great to merge this unless there are any red flags.

LeeTibbert commented:
@ekrich Thank you for your hard work over months (years?) on this.  Thank you also
for addressing my concerns. 

The nature of signal handling is that it will require iteration and cut-&-fit evolution.
In particular, a loop through the `wait.h` methods/functions.  I concur that this
is a good baselevel and join in asking that it be merged.

ekrich commented:
I think it looks better now and I rebased.

WojciechMazur commented on posixlib/src/main/resources/scala-native/signal.c:
```diff
@@ -0,0 +1,89 @@
+#include <signal.h>
```
Can you please add the following header to this file, allowing it to be ignored on Windows. I believe we would receive CI errors otherwise. 

```suggestion
#if defined(__unix__) || defined(__unix) || defined(unix) ||                   \
    (defined(__APPLE__) && defined(__MACH__))
#include <signal.h>
```

ekrich replied:
Fixed.

WojciechMazur commented on posixlib/src/main/resources/scala-native/signal.c:
```diff
@@ -0,0 +1,89 @@
+int scalanative_si_asyncio() { return SI_ASYNCIO; }
+int scalanative_si_mesgq() { return SI_MESGQ; }
+
+// scala native helper functions
```
We can remove this comment, since there are no more helper functions here, also we need to add #endif for previous suggestion
```suggestion
#endif // is Unix or MacOS
```

ekrich replied:
Ok, this is fixed as well as the weird formatting is the latest commit so it should be ready now.

WojciechMazur reviewed (changes requested):
Sorry for the late review. In my opinion it's ready to be merged. Only two minor changes are needed: Add header preprocessor condition (otherwise it would fail in current CI) and to remove the spurious comment in the posix signal.c.
